samples/bluetooth/peripheral_with_multiple_identities/src/main.c

```diff
@@ -0,0 +1,179 @@
+#include <stdio.h>
+#include <string.h>
+#include <zephyr/kernel.h>
+#include <zephyr/sys/printk.h>
+
+#include <zephyr/bluetooth/bluetooth.h>
+#include <zephyr/bluetooth/conn.h>
+
+#define DEVICE_NAME_MAX_INDEX_SIZE 4 /* The max size of the index added to the device name. */
+
+/* Structure to hold information about each advertiser. */
+struct advertiser_info {
+	struct k_work work;           /* Work item for the workqueue */
+	struct bt_le_ext_adv *adv;    /* Pointer to the advertising set */
+	uint8_t id;                   /* ID associated with the advertiser */
+};
+
+/* Buffer for dynamically constructing the device name to differentiate between advertisers. */
+static char device_name_buf[sizeof(CONFIG_BT_DEVICE_NAME) + DEVICE_NAME_MAX_INDEX_SIZE];
+
+static struct advertiser_info advertisers[CONFIG_BT_EXT_ADV_MAX_ADV_SET];
+
+static void start_connectable_advertiser(struct k_work *work);
+
+static void connected(struct bt_conn *conn, uint8_t err)
+{
+	char addr_str[BT_ADDR_LE_STR_LEN];
+
+	bt_addr_le_to_str(bt_conn_get_dst(conn), addr_str, sizeof(addr_str));
+	printk("connected to %s\n", addr_str);
+}
+
+static void disconnected(struct bt_conn *conn, uint8_t reason)
+{
+	char addr_str[BT_ADDR_LE_STR_LEN];
+
+	bt_addr_le_to_str(bt_conn_get_dst(conn), addr_str, sizeof(addr_str));
+
+	printk("disconnected %s, reason %u\n", addr_str, reason);
+
+	struct bt_conn_info connection_info;
+	int err;
+	err = bt_conn_get_info(conn, &connection_info);
+
+	if (err) {
+		printk("Failed to get conn info (err %d)\n", err);
+		return;
+	}
+
+	/* Get the ID of the disconnected advertiser. */
+	uint8_t id_current = connection_info.id;
+	printk("Advertiser %d disconnected\n", id_current);
+
+	/* Restart the advertiser by submitting its work to the workqueue. */
+	k_work_submit(&advertisers[id_current].work);
```

Comment:
I am not sure I understand entirely. Internally, in zephyrs id.c it seems to do a similar thing as the array holding the IDs is set to have a size of CONFIG_BT_ID_MAX in hci_core.h. In the id.c file it seems, like similar logic where the id is the index for the address array and it is between [0, CONFIG_BT_ID_MAX]. As we populated the array with the IDs being the index in the array I think this is safe to do. I can add a check ensuring that logically the connection_info.id returned is not larger than CONFIG_BT_ID_MAX.

Reply:
Yeah, you are right. Internally it is implemented like that, but the API documentation only states
```
 * @return Identity identifier (>= 0) in case of success, or a negative
 *         error code on failure.
```.
That is, they are allowed to change what range of values are returned.

Reply:
Decided offline to use an assert to enforce this behavior. Marking as resolved.